README
NAME
    DBIx::Class - Extensible and flexible object <-> relational mapper.

SYNOPSIS
DESCRIPTION
    This is an SQL to OO mapper, inspired by the Class::DBI framework, and
    meant to support compability with it, while restructuring the internals
    and making it possible to support some new features like self-joins,
    distinct, group bys and more.

    This project is still at an early stage, so the maintainers don't make
    any absolute promise that full backwards-compatibility will be
    supported; however, if we can without compromising the improvements
    we're trying to make, we will, and any non-compatible changes will merit
    a full justification on the mailing list and a CPAN developer release
    for people to test against.

    The community can be found via -

      Mailing list: http://lists.rawmode.org/mailman/listinfo/dbix-class/

      SVN: http://dev.catalyst.perl.org/repos/bast/trunk/DBIx-Class/

      Wiki: http://dbix-class.shadowcatsystems.co.uk/

      IRC: irc.perl.org#dbix-class

QUICKSTART
    If you're using Class::DBI, and want an easy and fast way of migrating
    to DBIx::Class, take a look at DBIx::Class::CDBICompat.

    There are two ways of using DBIx::Class, the "simple" way and the
    "schema" way. The "simple" way of using DBIx::Class needs less classes
    than the "schema" way but doesn't give you the ability to easily use
    different database connections.

    Some examples where different database connections are useful are:

    different users with different rights different databases with the same
    schema.

  Simple
    First you need to create a base class which all other classes will
    inherit from. See DBIx::Class::DB for information on how to do this.

    Then you need to create a class for every table you want to use with
    DBIx::Class. See DBIx::Class::Table for information on how to do this.

  Schema
    With this approach, the table classes inherit directly from
    DBIx::Class::Core, although it might be a good idea to create a "parent"
    class for all table classes that inherits from DBIx::Class::Core and
    adds additional methods needed by all table classes, e.g. reading a
    config file or loading auto primary key support.

    Look at DBIx::Class::Schema for information on how to do this.

    If you need more help, check out the introduction in the manual below.

SEE ALSO
    DBIx::Class::Core - DBIC Core Classes
    DBIx::Class::Manual - User's manual
    DBIx::Class::CDBICompat - Class::DBI Compat layer
    DBIx::Class::Schema
    DBIx::Class::ResultSet
    DBIx::Class::ResultSource
    DBIx::Class::Row - row-level methods
    DBIx::Class::PK - primary key methods
    DBIx::Class::Relationship - relationships between tables

AUTHOR
    Matt S. Trout <mst@shadowcatsystems.co.uk>

CONTRIBUTORS
    Andy Grundman <andy@hybridized.org>

    Brian Cassidy <bricas@cpan.org>

    Dan Kubb <dan.kubb-cpan@onautopilot.com>

    Dan Sully <daniel@cpan.org>

    David Kamholz <dkamholz@cpan.org>

    Jules Bean

    Marcus Ramberg <mramberg@cpan.org>

    Paul Makepeace

    CL Kao

    Jess Robinson

    Marcus Ramberg

    Will Hawes

    Todd Lipcon

    Daniel Westermann-Clark <danieltwc@cpan.org>

    Alexander Hartmaier <alex_hartmaier@hotmail.com>

    Zbigniew Lukasiak

    Nigel Metheringham <nigelm@cpan.org>

    Jesper Krogh

    Brandon Black

    Christopher H. Laco

    Scotty Allen <scotty@scottyallen.com>

    sc_

    Robert Sedlacek <phaylon@dunkelheit.at>

    Justin Guenther <jguenther@agr.gc.ca>

    Daisuke Murase <typester@cpan.org>

    Scott McWhirter (konobi)

    Cory Watson (gphat)

LICENSE
    You may distribute this code under the same terms as Perl itself.



Changes
Revision history for DBIx::Class

0.082843 2022-05-17 10:50 (UTC)
    * Fixes
        - Adjust tests to account for DBD::SQLite's column_info() no longer
          being case-preserving in recent versions
        - Fix t/storage/replicated.t relying on no-longer-available module
        - Adjust t/53lean_startup.t to work around spurious changes in
          Perl5 core (GH#143)

0.082842 2020-06-16 20:10 (UTC)
    * New Features
        - An on_connect rebase_sqlmaker call allowing experimentation with
          non-core SQL generators on a per-$schema-instance basis
          https://is.gd/DBIC_rebase_sqlmaker
        - Automatically detect and use multi-column IN on recent versions of
          libsqlite: ... WHERE ( foo, bar ) IN ( SELECT foo, bar FROM ... )

    * Fixes
        - Fix silent failure to retrieve a primary key (RT#80283) or worse:
          returning an incorrect value (RT#115381) in case a rdbms-side autoinc
          column is declared as PK with the is_auto_increment attribute unset
        - Fix overly-aggressive condition unrolling, corrupting custom ops with
          array arguments (RT#132390)
        - Fix docs on how to properly use Moo(se) in ResultSet's, and fix a
          corner case of ->count not functioning properly when the old recipe
          was being used (GH#105)
        - Fix incorrect ::Storage->_ping() behavior under Sybase (RT#114214)
        - Work around breakage in Hash::Merge by soft-requiring Clone as part
          of the replicated subsystem (RT#124321)

    * Misc
        - DBIC_TRACE_PROFILE=... now uses a ::Storage::Statistics subclass
          DBIx::Class::Storage::Debug::PrettyTrace which properly ships as a
          part of this distrinbution
        - Switch out SQL::Abstract dependency with a slower moving dist
        - Remove Data::Page dependency by inlining its entirety into the core
          DBIx::Class::ResultSet::Pager (RT#130686)

0.082841 2018-01-29 08:10 (UTC)
    * Test-suite fixup changes only - no reason to upgrade, wait for 0.082850

    * Misc
        - Unblock DBD::SQLite (RT#118395)
        - Fix missing ORDER BY leading to failures of t/prefetch/grouped.t
          under upcoming libsqlite (RT#117271)
        - Temporarily disable a non-critical portion of test failing on
          upcoming libsqlite, proper fix coming in 0.082850 (RT#119845)
        - Add test workaround for ( largely irrelevant ) RT#120129
        - Add temporary test-workaround for a devrel of DBD::SQLite
          RT#124227 / https://twitter.com/dbix_class/status/957271153751527424
        - Add workaround for one of the most damaging "improvements" made
          during the modern-perl-era (RT#120827)
        - Fix malformed README file encoding (RT#122028)

0.082840 2016-06-20 07:02 (UTC)
    * New Features
        - When using non-scalars (e.g. arrays) as literal bind values it is no
          longer necessary to explicitly specify a bindtype (this turned out
          to be a mostly useless overprotection)

    * Fixes
        - Ensure leaving an exception stack via Return::MultiLevel or something
          similar produces a large warning
        - Another relatively invasive set of ::FilterColumn changes, covering
          potential data loss (RT#111567). Please run your regression tests!
        - Ensure failing on_connect* / on_disconnect* are dealt with properly,
          notably on_connect* failures now properly abort the entire connect
        - Fix use of ::Schema::Versioned combined with a user-supplied
          $dbh->{HandleError} (GH#101)
        - Fix parsing of DSNs containing driver arguments (GH#99)
        - Fix silencing of exceptions thrown by custom inflate_result() methods
        - Fix complex prefetch when ordering over foreign boolean columns
          ( Pg can't MAX(boolcol) despite being able to ORDER BY boolcol )
        - Fix infinite loop on ->svp_release("nonexistent_savepoint") (GH#97)
        - Fix spurious ROLLBACK statements when a TxnScopeGuard fails a commit
          of a transaction with deferred FK checks: a guard is now inactivated
          immediately before the commit is attempted (RT#107159)
        - Fix the Sybase ASE storage incorrectly attempting to retrieve an
          autoinc value when inserting rows containing blobs (GH#82)
        - Remove spurious exception warping in ::Replicated::execute_reliably
          (RT#113339)
        - Work around unreliable $sth->finish() on INSERT ... RETURNING within
          DBD::Firebird on some compiler/driver combinations (RT#110979)
        - Fix leaktest failures with upcoming version of Sub::Quote
        - Really fix savepoint rollbacks on older DBD::SQLite (fix in 0.082800
          was not sufficient to cover up RT#67843)

    * Misc
        - Test suite is now officially certified to work under very high random
          parallelism: META x_parallel_test_certified set to true accordingly
        - Typo fixes from downstream debian packagers (RT#112007)

0.082821 2016-02-11 17:58 (UTC)
    * Fixes
        - Fix t/52leaks.t failures on compilerless systems (RT#104429)
        - Fix t/storage/quote_names.t failures on systems with specified Oracle
          test credentials while missing the optional Math::Base36
        - Fix test failures when DBICTEST_SYBASE_DSN is set (unnoticed change
          in error message wording during 0.082800 and a bogus test)
        - Remove largely obsolete test of SQLite view deployment (RT#111916)

    * Misc
        - Work around rare test deadlock under heavy parallelism (RT#108390)

0.082820 2015-03-20 20:35 (UTC)
    * Fixes
        - Protect destructors from rare but possible double execution, and
          loudly warn the user whenever the problem is encountered (GH#63)
        - Relax the 'self_result_object' argument check in the relationship
          resolution codepath, restoring exotic uses of inflate_result
          http://lists.scsys.co.uk/pipermail/dbix-class/2015-January/011876.html

Makefile.PL
use strict;
use warnings;

use 5.008001;


# Work around one of the most damaging "improvements" made
# during the modern-perl-era
BEGIN { push @INC, '.' unless grep { $_ eq '.' } @INC }


# For the time being simply generate UNKNOWN cperl test reports
# I would *LOVE* to be able to support it seamlessly, but with
# rurban adamant that CPAN ought to be gentrified... pity.
# https://irclog.perlgeek.de/perl11/2016-06-08#i_12624929 ( 06:57 ~ 09:34 )
BEGIN {
  die <<'EOE' if $^V =~ /c/;

Currently DBIx::Class is not attempting to be compatible with CPerl
For more info: https://irclog.perlgeek.de/perl11/2016-06-08#i_12624929

EOE
}


use inc::Module::Install 1.06;
BEGIN { makemaker_args( NORECURS => 1 ) } # needs to happen early for old EUMM

##
## DO NOT USE THIS HACK IN YOUR DISTS!!! (it makes #toolchain sad)
##
# get cpanX --installdeps . to behave in a checkout (most users do not expect
# the deps for a full test suite run, and if they do - there's MI::AutoInstall
# for that)
BEGIN {
  $Module::Install::AUTHOR = 0 if (grep { $ENV{"PERL5_${_}_IS_RUNNING"} } (qw/CPANM CPANPLUS CPAN/) );
}

name         'DBIx-Class';
version_from 'lib/DBIx/Class.pm';
perl_version '5.008001';

###
### DO NOT ADD OPTIONAL DEPENDENCIES HERE, EVEN AS recommends()
### All of them *MUST* go to DBIx::Class::Optional::Dependencies
###
my $runtime_requires = {

  # DBI itself should be capable of installation and execution in pure-perl
  # mode. However it has never been tested yet, so consider XS for the time
  # being
###
### IMPORTANT - do not raise this dependency
### even though many bugfixes are present in newer versions, the general DBIC
### rule is to bend over backwards for available DBI versions (given upgrading
### them is often *not* easy or even possible)
###
  'DBI'                      => '1.57',

  # XS (or XS-dependent) libs
  'Sub::Name'                => '0.04',

  # pure-perl (FatPack-able) libs
  'Class::Accessor::Grouped' => '0.10012',
  'Class::C3::Componentised' => '1.0009',
  'Class::Inspector'         => '1.24',
  'Config::Any'              => '0.20',
  'Context::Preserve'        => '0.01',
  'Data::Dumper::Concise'    => '2.020',
  'Devel::GlobalDestruction' => '0.09',
  'Hash::Merge'              => '0.12',
  'Moo'                      => '2.000',
  'MRO::Compat'              => '0.12',
  'Module::Find'             => '0.07',
  'namespace::clean'         => '0.24',
  'Path::Class'              => '0.18',
  'Scope::Guard'             => '0.03',
  'SQL::Abstract::Classic'   => '1.91',
  'Try::Tiny'                => '0.07',

  # Technically this is not a core dependency - it is only required
  # by the MySQL codepath. However this particular version is bundled
  # since 5.10.0 and is a pure-perl module anyway - let it slide
  'Text::Balanced'           => '2.00',
};

my $build_requires = {
};

my $test_requires = {
  'File::Temp'               => '0.22',
  'Test::Deep'               => '0.101',
  'Test::Exception'          => '0.31',
  'Test::Warn'               => '0.21',
  'Test::More'               => '0.94',

  # needed for testing only, not for operation
  # we will move away from this dep eventually, perhaps to DBD::CSV or something
###
### IMPORTANT - do not raise this dependency
### even though many bugfixes are present in newer versions, the general DBIC
### rule is to bend over backwards for available DBDs (given upgrading them is
### often *not* easy or even possible)
###
  'DBD::SQLite'              => '1.29',

  # this is already a dep of n::c, but just in case - used by t/55namespaces_cleaned.t
  # remove and do a manual glob-collection if n::c is no longer a dep
  'Package::Stash'           => '0.28',
};

# if the user has this env var set and no SQLT installed, tests will fail
# Note - this is added as test_requires *directly*, so it gets properly
# excluded on META.yml cleansing (even though no dist can be created from this)
# we force this req regarless of author_deps, worst case scenario it will
# be specified twice
#
# also note that we *do* set dynamic_config => 0, as this is the only thing
# that we determine dynamically, and in all fairness if someone sets the
# envvar *and* is not running a full Makefile/make/maketest cycle - they get
# to keep the pieces
if ($ENV{DBICTEST_SQLT_DEPLOY}) {
  local @INC = ('lib', @INC);
  require DBIx::Class::Optional::Dependencies;
  my $dep_req = DBIx::Class::Optional::Dependencies->req_list_for('deploy');
  for (keys %$dep_req) {
    test_requires ($_ => $dep_req->{$_})
  }
}

tests_recursive ('t');
tests_recursive ('xt') if (
  $Module::Install::AUTHOR
    or
  $ENV{DBICTEST_RUN_ALL_TESTS}
    or
  (
    ( $ENV{TRAVIS}||'' ) eq 'true'
      and
    ($ENV{TRAVIS_REPO_SLUG}||'') =~ m|\w+/DBIx-Class$|
  )
    or
  ( $ENV{AUTOMATED_TESTING} and ! $ENV{PERL5_CPANM_IS_RUNNING} and ! $ENV{RELEASE_TESTING} )
);

install_script (qw|
    script/dbicadmin
|);

# this is so we can order requires alphabetically
# copies are needed for potential author requires injection
my $reqs = {
  build_requires => { %$build_requires },
  requires => { %$runtime_requires },
  test_requires => { %$test_requires },
};

# only do author-includes if not part of a `make` run
if ($Module::Install::AUTHOR  and ! $ENV{MAKELEVEL}) {
  invoke_author_mode()
}
else {
  # make sure this Makefile can not be used to make a dist
  # (without the author includes there are no meta cleanup, no sanity checks, etc)
  postamble <<EOP;
create_distdir: nonauthor_stop_distdir_creation
nonauthor_stop_distdir_creation:
\t\$(NOECHO) \$(ECHO) Creation of dists in non-author mode is not allowed
\t\$(NOECHO) \$(FALSE)
EOP
}

# compose final req list, for alphabetical ordering
my %final_req;
for my $rtype (keys %$reqs) {
  for my $mod (keys %{$reqs->{$rtype}} ) {

    # sanity check req duplications
    die "$mod specified as both a '$rtype' and a '$final_req{$mod}[0]'\n"
      if $final_req{$mod};

    $final_req{$mod} = [ $rtype, $reqs->{$rtype}{$mod}||0 ],
  }
}

# actual require
for my $mod (sort keys %final_req) {
  my ($rtype, $ver) = @{$final_req{$mod}};
  no strict 'refs';
  $rtype->($mod, $ver);
}

# author-mode or not - this is where we show a list of missing deps
# IFF we are running interactively
auto_install();

{
  # M::I understands unicode in meta but does not write with the right
  # layers - fhtagn!!!
  local $SIG{__WARN__} = sub { warn $_[0] unless $_[0] =~ /Wide character in print/ };
  WriteAll();
}

exit 0;


###
### Nothing user-serviceable beyond this point
### (none of this executes on regular install)
###


# needs to be here to keep 5.8 string eval happy
# (the include of Makefile.PL.inc loop)
my $mm_proto;

sub invoke_author_mode {
  # get options here, make $args available to all snippets
  require Getopt::Long;
  my $getopt = Getopt::Long::Parser->new(
    config => [qw/gnu_getopt bundling_override no_ignore_case pass_through/]
  );
  my $args = {
    skip_author_deps => undef,
  };
  $getopt->getoptions($args, qw/
    skip_author_deps|skip-author-deps
  /);
  if (@ARGV) {
    warn "\nIgnoring unrecognized option(s): @ARGV\n\n";
  }

  # We need the MM facilities to generate the pieces for the final MM run.
  # Just instantiate a throaway object here
  #
  # Also EUMM and MI disagree on what is the format of Meta->name, just
  # punt here until a new M::I is shipped (if at all)
  my $name = Meta->name || die 'The Module::Install metadata must be available at this point but is not - did you rearrange the Makefile.PL...?';
  $name =~ s/\-/::/g;
  $mm_proto = ExtUtils::MakeMaker->new({
    NORECURS => 1,
    NAME => $name,
  });

  # Crutch for DISTBUILDING_IN_HELL
  # Spits back a working dos2unix snippet to be used on the supplied path(s)
  # Ironically EUMM's dos2unix is broken on win32 itself - it does
  # not take into account the CRLF layer present on win32
  my $crlf_fixup = sub {
    return '' unless ($^O eq 'MSWin32' or $^O eq 'cygwin');
    my $targets = join ', ', map { "q($_)" } @_;
    "\t" . $mm_proto->oneliner( qq(\$ENV{PERLIO}='unix' and system( \$^X, qw( -MExtUtils::Command -e dos2unix -- ), $targets ) ) );
  };

  # we are in the process of (re)writing the makefile - some things we
  # call below very well may fail
  local $ENV{DBICTEST_NO_MAKEFILE_VERIFICATION} = 1;

  require File::Spec;
  # string-eval, not do(), because we need to provide the
  # $mm_proto, $reqs and $*_requires lexicals to the included file
  # (some includes *do* modify $reqs above)
  for my $inc (sort glob ( File::Spec->catfile('maint', 'Makefile.PL.inc', '*') ) ) {
    my $src = do { local (@ARGV, $/) = $inc; <> } or die $!;
    eval "use warnings; use strict; $src" or die sprintf
      "Failed execution of %s: %s\n",
      $inc,
      ($@ || $! || 'Unknown error'),
    ;
  }
}
